### PR TITLE
[JENKINS-14572] Add "Do not fetch tags" advanced clone option

### PR DESCRIPTION
**Pull request by rhaendel**
As discussed in [JENKINS-14572](https://issues.jenkins-ci.org/browse/JENKINS-14572), fetching all tags is the git plugin's default behaviour - even when making shallow clones using the appropriate advanced clone behaviour.

I've added a check box "Do not fetch tags" to the advanced clone behaviours. This way a user can explicitly override the default.

### CODE REVIEW

File: src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java [hunk 1]
```diff
@@ -20,13 +20,19 @@
  * @author Kohsuke Kawaguchi
  */
 public class CloneOption extends GitSCMExtension {
-    private boolean shallow;
-    private String reference;
-    private Integer timeout;
+    private final boolean shallow;
```
Comment (KostyaSha) [inline src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java]: why are you making them final?

File: src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java [hunk 2]
```diff
@@ -49,13 +59,18 @@ public void decorateCloneCommand(GitSCM scm, Run<?, ?> build, GitClient git, Tas
             listener.getLogger().println("Using shallow clone");
             cmd.shallow();
         }
+        if (noTags) {
```
Comment (KostyaSha) [inline src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java]: `if (isNoTags())`

File: src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java [hunk 3]
```diff
@@ -49,13 +59,18 @@ public void decorateCloneCommand(GitSCM scm, Run<?, ?> build, GitClient git, Tas
...
+            cmd.tags(false);
+        }
         cmd.timeout(timeout);
         cmd.reference(build.getEnvironment(listener).expand(reference));
     }
     
     @Override
     public void decorateFetchCommand(GitSCM scm, GitClient git, TaskListener listener, FetchCommand cmd) throws IOException, InterruptedException, GitException {
-    	cmd.timeout(timeout);
+        cmd.tags(!noTags);
```
Comment (KostyaSha) [inline src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java]: why indentation changed?

Comment (daniel-beck) [inline src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java]: It was three spaces, now it's four, and right next to a modified line. Is that really how strictly you enforce "no unrelated formatting changes"?

Comment (KostyaSha) [inline src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java]: Just found thus asked

Comment (KostyaSha) [inline src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java]: Btw in email indentation is different.

Comment (rhaendel) [inline src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java]: What GitHub's code code viewer doesn't show here: It were four spaces followed by one tab. The whole file is formatted using spaces, except this one line. I had to add a line right next to it. So I decided to correct it and replaced the tab with four spaces.

My interpretation of "no unrelated formatting changes" was that I should not use auto formatters that change the format all over the file which obfuscates the actual change and makes merging other pull requests touching the same file difficult.

File: src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java (hunk 2 above)
Comment (rhaendel) [inline src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java]: Technically, I agree.
But I tried to follow the code style that was already there. In the whole class, private fields are accessed directly and not by using the public accessor methods. Just take a look at the preceding four lines or the next four lines to see what I mean.

File: src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java (hunk 1 above)
Comment (rhaendel) [inline src/main/java/hudson/plugins/git/extensions/impl/CloneOption.java]: CloneOption is used as a [value object](http://martinfowler.com/bliki/ValueObject.html). Its attributes are set once by the constructor (or the object is deserialized from xml file) and never changed again.
Making them final (and CloneOption immutable) makes reading and understanding the code easier. You simply don't have to think about what happens if one of these values changes (especially in a multi threading environment). It just doesn't happen. The compiler checks it for you.

**Comment by rhaendel** [comment]
:+1:

**Comment by Gamesh** [comment]
when will this get released ?

**Comment by MarkEWaite** [comment]
Was released two years ago

**Comment by Gamesh** [comment]
but git still fetches all tags and there is no option to disable it, am i missing something ?

**Comment by MarkEWaite** [comment]
As far as I know,  when the advanced clone options is added,  and the "No
tags" box is checked, the tags are not fetched.  I won't bee able to test
that myself today,  too many other things to do.

**Comment by Gamesh** [comment]
OK found it, thanks
